Build issue

**Describe the bug**

I am new to Kaoto.io, and I am trying to run the build as per the Readme.md file, however I keep getting errors.

**To Reproduce**
Steps to reproduce the behavior:
1. Clone Repo (tried both main branch & release)
2. nvm install (tried both on parent, and inside each module)

**Expected behavior**
Be able to build and run the code as per Readme.MD

**Logs**

[INFO] Reactor Summary for parent 0.7.6:
[INFO]
[INFO] parent ............................................. SUCCESS [  2.015 s]
[INFO] model .............................................. SUCCESS [ 10.345 s]
[INFO] metadata ........................................... SUCCESS [ 31.050 s]
[INFO] catalog ............................................ SUCCESS [  6.985 s]
[INFO] services-interfaces ................................ SUCCESS [  5.829 s]
[INFO] kamelet-support .................................... FAILURE [01:29 min]
[INFO] camel-route-support ................................ SKIPPED
[INFO] cluster ............................................ SKIPPED
[INFO] api ................................................ SKIPPED

`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0:test (default-test) on project kamelet-support: There are test failures.
[ERROR]
[ERROR] Please refer to C:\Users\mz\Downloads\kaoto-backend-0.7.6\kaoto-backend-0.7.6\kamelet-support\target\surefire-reports for the individual test results.`

**Environment **
Windows 10
Maven home: C:\Program Files\Maven\apache-maven-3.9.1
Java version: 17.0.6, vendor: Oracle Corporation, runtime: C:\Program Files\Java\jdk-17
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

** I tried**
mvn dependency:resolve -U   <- did not work

It is failing for me too.
```
[INFO] Running io.kaoto.backend.model.deployment.kamelet.step.TryCatchTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 s - in io.kaoto.backend.model.deployment.kamelet.step.TryCatchTest
2023-03-31 15:30:50,939 WARN  [io.qua.ope.run.exp.otl.LateBoundBatchSpanProcessor] (main) No BatchSpanProcessor delegate specified, no action taken.
2023-03-31 15:30:51,018 INFO  [io.quarkus] (main) kamelet-support stopped in 0.085s
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   KameletStepParserServiceTest.checkEIP:218 
...
```

--- a/camel-route-support/src/test/java/io/kaoto/backend/api/service/step/parser/camelroute/CamelRouteStepParserServiceTest.java
+++ b/camel-route-support/src/test/java/io/kaoto/backend/api/service/step/parser/camelroute/CamelRouteStepParserServiceTest.java
@@ -68,7 +68,8 @@ class CamelRouteStepParserServiceTest {
 
     @ParameterizedTest
     @ValueSource(strings = {"route.yaml", "route2-complex-expressions.yaml",
-            "route3-complex-expressions.yaml", "route-ids.yaml", "route4-pathparams.yaml", "route5-placeholders.yaml"})
+            "route3-complex-expressions.yaml", "route-ids.yaml", "route4-pathparams.yaml", "route5-placeholders.yaml",
+            "route6-un-marshal.yaml"})
     void deepParseParametrized(String file) throws IOException {
         var route = new String(this.getClass().getResourceAsStream(file).readAllBytes(),
                 StandardCharsets.UTF_8);

--- a/camel-route-support/src/test/resources/io/kaoto/backend/api/service/step/parser/camelroute/route6-un-marshal.yaml
+++ b/camel-route-support/src/test/resources/io/kaoto/backend/api/service/step/parser/camelroute/route6-un-marshal.yaml
@@ -1,0 +1,15 @@
+- from:
+    uri: timer:null
+    steps:
+    - marshal:
+        data-format:
+          json:
+            library: Gson
+    - unmarshal:
+        data-format:
+          json:
+            property: something
+            another: whatever
+    - unmarshal:
+        xml:
+          class: classInJava.class

--- a/kamelet-support/src/main/java/io/kaoto/backend/KamelPopulator.java
+++ b/kamelet-support/src/main/java/io/kaoto/backend/KamelPopulator.java
@@ -63,7 +63,6 @@ import io.kaoto.backend.model.deployment.kamelet.step.UriFlowStep;
 import io.kaoto.backend.model.deployment.kamelet.step.ValidateFlowStep;
 import io.kaoto.backend.model.deployment.kamelet.step.WireTapFlowStep;
 import io.kaoto.backend.model.deployment.kamelet.step.dataformat.DataFormat;
-import io.kaoto.backend.model.parameter.ArrayParameter;
 import io.kaoto.backend.model.parameter.Parameter;
 import io.kaoto.backend.model.parameter.StringParameter;
 import io.kaoto.backend.model.step.Branch;
@@ -559,24 +558,9 @@ public class KamelPopulator {
     }
 
     private void assignParameters(final Step step, final MarshalFlowStep marshal) {
-        marshal.setDataFormat(new DataFormat());
-        marshal.getDataFormat().setProperties(new HashMap<>());
-
         for (Parameter p : step.getParameters()) {
-            if (null == p.getValue()) {
-                continue;
-            }
-
-            if ("dataformat".equalsIgnoreCase(p.getId())) {
-                marshal.getDataFormat().setFormat(p.getValue().toString());
-            } else if ("properties".equalsIgnoreCase(p.getId())) {
-                for (var param : ((ArrayParameter) p).getValue()) {
-                    if (param instanceof Object[] array) {
-                        marshal.getDataFormat().getProperties().put(array[0].toString(), array[1].toString());
-                    } else if (param instanceof List list) {
-                        marshal.getDataFormat().getProperties().put(list.get(0).toString(), list.get(1).toString());
-                    }
-                }
+            if (null != p.getValue() && "dataformat".equalsIgnoreCase(p.getId())) {
+                marshal.setDataFormat(new DataFormat(p.getValue()));
             }
         }
     }

--- a/kamelet-support/src/main/java/io/kaoto/backend/model/deployment/kamelet/step/MarshalFlowStep.java
+++ b/kamelet-support/src/main/java/io/kaoto/backend/model/deployment/kamelet/step/MarshalFlowStep.java
@@ -30,7 +30,9 @@ public class MarshalFlowStep implements FlowStep {
         Map<String, Object> step = new HashMap<>();
         Map<String, Object> properties = new HashMap<>();
         step.put("marshal", properties);
-        properties.put(dataFormat.getFormat(), dataFormat.getProperties());
+        if (dataFormat != null) {
+            properties.put(dataFormat.getFormat(), dataFormat.getProperties());
+        }
         return step;
     }
 
@@ -48,15 +50,7 @@ public class MarshalFlowStep implements FlowStep {
     protected void assignParameters(final Step res) {
         for (var param : res.getParameters()) {
             if ("dataformat".equalsIgnoreCase(param.getId())) {
-                param.setValue(this.getDataFormat().getFormat());
-            } else if ("properties".equalsIgnoreCase(param.getId())
-                            && this.getDataFormat().getProperties() != null) {
-                Object[] array = new Object[this.getDataFormat().getProperties().size()];
-                int i = 0;
-                for (var entry : this.getDataFormat().getProperties().entrySet()) {
-                    array[i++] = new Object[]{entry.getKey(), entry.getValue()};
-                }
-                param.setValue(array);
+                param.setValue(this.getDataFormat());
             }
         }
     }
@@ -66,8 +60,7 @@ public class MarshalFlowStep implements FlowStep {
         return dataFormat;
     }
 
-    public void setDataFormat(
-            final DataFormat dataFormat) {
+    public void setDataFormat(final DataFormat dataFormat) {
         this.dataFormat = dataFormat;
     }
 }

--- a/kamelet-support/src/main/java/io/kaoto/backend/model/deployment/kamelet/step/UnmarshalFlowStep.java
+++ b/kamelet-support/src/main/java/io/kaoto/backend/model/deployment/kamelet/step/UnmarshalFlowStep.java
@@ -10,9 +10,7 @@ import java.io.Serial;
 import java.util.HashMap;
 import java.util.Map;
 
-@JsonDeserialize(
-        using = MarshalDeserializer.class
-)
+@JsonDeserialize(using = MarshalDeserializer.class)
 public class UnmarshalFlowStep extends MarshalFlowStep {
     @Serial
     private static final long serialVersionUID = 5841231681362382129L;
@@ -22,7 +20,9 @@ public class UnmarshalFlowStep extends MarshalFlowStep {
         Map<String, Object> step = new HashMap<>();
         Map<String, Object> properties = new HashMap<>();
         step.put("unmarshal", properties);
-        properties.put(getDataFormat().getFormat(), getDataFormat().getProperties());
+        if (getDataFormat() != null) {
+            properties.put(getDataFormat().getFormat(), getDataFormat().getProperties());
+        }
         return step;
     }
 

--- a/kamelet-support/src/main/java/io/kaoto/backend/model/deployment/kamelet/step/dataformat/DataFormat.java
+++ b/kamelet-support/src/main/java/io/kaoto/backend/model/deployment/kamelet/step/dataformat/DataFormat.java
@@ -2,6 +2,7 @@ package io.kaoto.backend.model.deployment.kamelet.step.dataformat;
 
 import java.io.Serial;
 import java.io.Serializable;
+import java.util.HashMap;
 import java.util.Map;
 
 public class DataFormat implements Serializable {
@@ -10,7 +11,26 @@ public class DataFormat implements Serializable {
     private static final long serialVersionUID = 239784528129L;
 
     private String format;
-    private Map<String, String> properties;
+    private Map<String, Object> properties;
+
+    public DataFormat() {
+        //Needed for serialization
+    }
+
+    public DataFormat(Object dataformat) {
+        this.setProperties(new HashMap<>());
+        if (dataformat instanceof DataFormat df) {
+            this.setFormat(df.getFormat());
+            this.setProperties(df.getProperties());
+        } else if (dataformat instanceof Map map) {
+            this.setFormat(String.valueOf(map.getOrDefault("format", "")));
+
+            final var potentialProperties = map.getOrDefault("properties", null);
+            if (potentialProperties instanceof Map) {
+                this.setProperties((Map<String, Object>) potentialProperties);
+            }
+        }
+    }
 
     public String getFormat() {
         return format;
@@ -20,12 +40,11 @@ public class DataFormat implements Serializable {
         this.format = format;
     }
 
-    public Map<String, String> getProperties() {
+    public Map<String, Object> getProperties() {
         return properties;
     }
 
-    public void setProperties(
-            final Map<String, String> properties) {
+    public void setProperties(final Map<String, Object> properties) {
         this.properties = properties;
     }
 }

--- a/kamelet-support/src/main/java/io/kaoto/backend/model/deployment/kamelet/step/marshal/MarshalDeserializer.java
+++ b/kamelet-support/src/main/java/io/kaoto/backend/model/deployment/kamelet/step/marshal/MarshalDeserializer.java
@@ -4,12 +4,14 @@ import com.fasterxml.jackson.core.JsonParser;
 import com.fasterxml.jackson.databind.DeserializationContext;
 import com.fasterxml.jackson.databind.JsonDeserializer;
 import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ValueNode;
 import io.kaoto.backend.model.deployment.kamelet.step.MarshalFlowStep;
 import io.kaoto.backend.model.deployment.kamelet.step.UnmarshalFlowStep;
 import io.kaoto.backend.model.deployment.kamelet.step.dataformat.DataFormat;
 import org.jboss.logging.Logger;
 
-import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
 
 public class MarshalDeserializer extends JsonDeserializer {
     private final Logger log = Logger.getLogger(MarshalDeserializer.class);
@@ -33,13 +35,8 @@ public class MarshalDeserializer extends JsonDeserializer {
 
                 var field = fields.next();
                 dataFormat.setFormat(field.getKey());
-                dataFormat.setProperties(new HashMap<>());
-                field.getValue().fields().forEachRemaining(e -> {
-                    dataFormat.getProperties().put(
-                            e.getKey(),
-                            e.getValue().asText());
-                });
-
+                dataFormat.setProperties(new LinkedHashMap<>());
+                extractProperties(dataFormat.getProperties(), field);
             }
             if (fields.hasNext()) {
                 log.error("Found a second data format on this marshal? "
@@ -51,5 +48,24 @@ public class MarshalDeserializer extends JsonDeserializer {
         }
 
         return step;
+    }
+
+    private static void extractProperties(final Map<String, Object> properties,
+                                          final Map.Entry<String, JsonNode> field) {
+        if (field.getValue() instanceof ValueNode v) {
+            properties.put(field.getKey(), v.asText());
+        } else {
+            field.getValue().fields().forEachRemaining(e -> {
+                if (e.getValue() instanceof ValueNode v) {
+                    properties.put(e.getKey(), v.asText());
+                } else {
+                    Map<String, Object> innerProperties = new LinkedHashMap<>();
+                    JsonNode node = e.getValue();
+                    node.fields().forEachRemaining(
+                            stringJsonNodeEntry -> extractProperties(innerProperties, stringJsonNodeEntry));
+                    properties.put(e.getKey(), innerProperties);
+                }
+            });
+        }
     }
 }

--- a/kamelet-support/src/main/java/io/kaoto/backend/model/deployment/kamelet/step/marshal/MarshalSerializer.java
+++ b/kamelet-support/src/main/java/io/kaoto/backend/model/deployment/kamelet/step/marshal/MarshalSerializer.java
@@ -33,7 +33,7 @@ public class MarshalSerializer extends JsonSerializer {
         if (df.getProperties() != null) {
             gen.writeStartObject();
             for (var property : df.getProperties().entrySet()) {
-                gen.writeStringField(property.getKey(), property.getValue());
+                gen.writeStringField(property.getKey(), property.getValue().toString());
             }
             gen.writeEndObject();
         }

